### DISCUSSION TITLE
Defining `core` composables for Modals

### DISCUSSION THREAD

**Discussion by onmax** [unovue/reka-ui#270] [General]
# What is this?

I have a proposal for the modal components (dialogue, alert, popover...). As this is a big change, I think it is better to discuss it first, improve it and then work on it.

> **Warning**
> The code provided in this discussion has not been extensively tested, but you can check out the idea and how it works in the `Popover` in this [commit](https://github.com/radix-vue/radix-vue/commit/6b63c9e8829ea89f84d9f9b19b1f096634afa23f) (which for now it works for the basic cases)

# Composables

## `useEmitCancelableEvent`

[`Implementation`](https://github.com/radix-vue/radix-vue/blob/e6f7155679b6e9a02532ac179430bac8625001e4/packages/radix-vue/src/shared/useEmitCancelableEvent.ts
)

We create an event and emit it. If the user calls `e.preventDefault()` on this event, the callback will not be called.

#### Usage

```ts
    useEmitCancelableEvent(emit, "pointer-down-outside", hidePopover, e);
    useEmitCancelableEvent(emit, "escape-key-down", hidePopover, e);
```

## `useEventListener`

[`Implementation`](https://github.com/radix-vue/radix-vue/blob/e6f7155679b6e9a02532ac179430bac8625001e4/packages/radix-vue/src/shared/useEventListener.ts)

This will allow us to listen to events. Additionally, you can add a flag called `disabled` which will automatically remove the listener when it is `false` and re-enable it when it is `true`.

#### Usage

```ts
const disabled = computed(() => !open.value);

useEventListener(
  document.documentElement,
  "click",
  (e) => {
    useEmitCancelableEvent(emit, "pointer-down-outside", hidePopover, e);
  },
  { disabled, exclude: [primitiveElement, triggerElement] } // The exlcude will ignore the event
);

useEventListener(
  document.documentElement,
  "keydown",
  (e) => {
    if (e.key !== "Escape") return;
    useEmitCancelableEvent(emit, "escape-key-down", hidePopover, e);
  },
  { disabled }
);
```

## `useModalInteraction`

[`Implementation`](https://github.com/radix-vue/radix-vue/blob/6b63c9e8829ea89f84d9f9b19b1f096634afa23f/packages/radix-vue/src/shared/useModalInteraction.ts)

It will handle the events that might occur when the modal opens or closes. Basically, the default behaviour with modals is:
1. When the modal **opens**: Automatically focus the user on any focusable element.
2. When the modal **closes**: Focus on the triggerable element.

Both of these events should be optional, and the user can cancel them. This composable basically handles that case and also locks the body if it is required.

Other options/features can be added to this composable whenever the modal is open/closed in the future

#### Usage

```ts
useModalInteraction(primitiveElement, { open, triggerElement }, emit);
```

## `useFocusOutside`

I haven't done anything related to this, but basically, detect when focus goes out and otherwise emit a cancelable event and close the modal.

##  `useInjected`

*Not related specifically to modals, but to the whole package.*

[`Implementation`](https://github.com/radix-vue/radix-vue/blob/6b63c9e8829ea89f84d9f9b19b1f096634afa23f/packages/radix-vue/src/shared/useInjected.ts)

Automatically inject values and return them as JSON. An error will be thrown if the key doesn't exist.

> This composable works super well, and will remove so many `injectedValue?.` and make the code more readable everywhere.

#### Usage
```ts
const { open, triggerElement, hidePopover, dataState, contentId } = useInjected(POPOVER_INJECTION_KEY);
```

**Comment by zernonia** [comment]
Nicely done on these composable! I have some comment here ya

1. In order for `useEmitCancelableEvent`, we need `await nextTicket()` for domain flush, then the `defaultPrevented` will be correct. Checkout [playground](https://play.vuejs.org/#eNrlVMFu2zAM/RXWF6dA52DorXOCrlmAbcDaYSuww7yD6tCOWlkyJDrNEOTfR8mJ58RdPmDNJREfH/VIPmUTva/rZNVgdBWlLreyJnBITT3NtKxqYwk2YLGALRTWVBBzatxBt8ZWQs1XqOmmITJ6l5SMe4ivHr/rOLPGkale4vSQPSfTudGOoHIlTLyOUfwRlTLww1i1OIvPM52OW9ksmA+EVa0EIZ8A0uXb6WYTyNttOuZTiA5VX+dK5k9JbdEHQxYAoSOpy0Dp99NywnWMDNu5zkPoDVlZlmgPqrYQhEBbeMDnvHTctRFdROR4BoUsk0dnNK9p44lZlJuqlgrtXU2SZ5RFVxAQjwke0fPnECPb4MU+ni8xf3oh/ujWPpZFXy06tCvMog4jYUukFp5/v8U1/+7AyiwaxdknwG/ojGq8xjbtptELlt3LC2o/BXfwvO/dfE2o3b4pL9RnbkN+FrEzZida/yv3MrkMvExveYpHhjy2OyihywnTuVrf+ppbumdvHPi/Z0ysJDm25gILqXHuT6Of8aED4l9s0z1hKXz7M+83pgn3W+cwwtUVBGnnMJnue/H5RmGiTDnKogcsjPV7YfMkfJtoFPG2PAkXXN8zxLOQ1EkehVuHlURBaOHD3RcoVOOWp0vKgsUNQRA8fuJiOwSkg53PcQErKcJgzviTRVyJd/DPZ5o+9N8gr6A3Id5F6pShKbPbrzZ58EaO/jr+y+0GMaPj2z3jla9/+wcWUllW) 
2. I think we can rename `useEmitCancelableEvent`, It's not quite explicit enough what it does 😬
3. I think it is better to have `DissmissableLayer` (similar to RadixUI) instead of this `useModalInteraction`, since there are many components is using that.

**Reply by onmax** [reply]
Thanks for you answer 👍 
1. I don't think we need to add `await nextTick()` for custom events.
The function flow is `event -> useEmitCancelableEvent -> emit event -> check if event has been prevented`
This is different to native events, and since out composable is only going to trigger custom events, I don't think we need it. Please correct me if I am wrong. 
Here the updated [Playground](https://play.vuejs.org/#eNrdVMFu2zAM/RXWF6dA52DorXOCrlmAbcDaYSuww7yD6tCOWlkyJDrNEOTfR8mJ58RtPqC5JOLjox7Jp2yij3WdrBqMrqLU5VbWBA6pqaeZllVtLMEGLBawhcKaCmJOjTvo1thKqPkKNd00REbvkpJxD/HV4w8dZ9Y4MtVLnB6y52Q6N9oRVK6Eidcxij+jUgZ+GasWZ/F5ptNxK5sF84GwqpUg5BNAunw/3WwCebtNx3wK0aHq61zJ/CmpLfpgyAIgdCR1GSj9flpOuI6RYTvXeQi9IyvLEu1B1RaCEGgLD/icl467NqKLiBzPoJBl8uiM5jVtPDGLclPVUqG9q0nyjLLoCgLiMcEjev4aYmQbvNjH8yXmTy/EH93ax7Lou0WHdoVZ1GEkbInUwvOft7jm3x1YmUWjOPsE+AOdUY3X2KbdNHrBsnt5Qe2X4A6e972brwm12zflhfrMbcjPInbG7ETr/+VeJpeBl+ktT/HIkMd2ByV0OWE6V+tbX3NL9+yNA//3jImVJMfWXGAhNc79afQ7PnRA/IdtuicshW9/5v3GNOH+6hxGuLqCIO0cJtN9Lz7fKEyUKUdZ9ICFsX4vbJ6EbxONIt6WJ+GC63uGeBaSOsmjcOuwkigILXy6+waFatzydElZsLghCILHT1xsh4B0sPM5LmAlRRjMGX+yiCvxDl59pulD/w3yCnoT4l2kThmaMrv9apMHb+Tor+NNbjeIGR3f7hm7hLe+5u0/kOlRcQ==)

2. I agree, I don't really like the current name, but out of ideas 🙈 
3. Sounds great. That composable is not really a composable, but not sure how to implement it. Would you mind to help with that?

**Reply by zernonia** [reply]
1. I dont think we should only consider triggering custom events, in `NavigationMenuLink`, it is a `@click` event where we need to be prevented. thus I think we should add `nextTick` to consider all cases.
2. Hahaha I can't think of a proper name too 😂
